src/directory/managed_directory.rs

```diff
@@ -207,26 +210,99 @@ impl ManagedDirectory {
         }
         Ok(())
     }
+
+    /// Verify checksum of a managed file
+    pub fn validate_checksum(&self, path: &Path) -> result::Result<bool, OpenReadError> {
+        let reader = self.directory.open_read(path)?;
+        let data = reader.as_slice();
+        let len = data.len() - CRC_SIZE;
+        let mut hasher = Hasher::new();
+        hasher.update(&data[..len]);
+        let crc = hasher.finalize().to_be_bytes();
+        Ok(data[len..]==crc)
+    }
+
+    /// List files for which checksum does not match content
+    pub fn list_damaged(&self) -> result::Result<HashSet<PathBuf>, OpenReadError> {
+        let mut hashset = HashSet::new();
+        let meta_informations_rlock = self
+            .meta_informations
+            .read()
+            .expect("Managed directory rlock poisoned in list damaged.");
+
+        println!("list_of_segments={:?}", meta_informations_rlock.managed_paths);
+        for path in &meta_informations_rlock.managed_paths {
+            if !self.validate_checksum(path)? {
+                hashset.insert(path.clone());
+            }
+        }
+        Ok(hashset)
+    }
+}
+
+struct CrcProxy<W: Write> {
+    hasher: Option<Hasher>,
+    writer: W,
+}
+
+impl<W: Write> CrcProxy<W> {
+    fn new(writer: W) -> Self {
+        CrcProxy {
+            hasher: Some(Hasher::new()),
+            writer
+        }
+    }
+}
+
+impl<W: Write> Drop for CrcProxy<W> {
+    fn drop(&mut self) {
+        // this should probably be logged
+        let _ = self.writer.write_all(&self.hasher.take().unwrap().finalize().to_be_bytes());
```

Comment:
IO in `drop` looks cool but it cannot throw.

Let's assume the fs fucked up on these last 4 bytes. (granted that's not a common event).
We want to detect the problem as soon as possible.
Also we rely on the idea that the underlying `io::Write` flushes on `Drop` which is not enforced by the trait I believe.

`.flush()` looks like a nicer place to put that in, but I assume you were put off by the possibility that tantivy (or someone), may call `.flush` more than once.

Having our own `Write` trait with a `.terminate(self) -> io::Result<()>` method would be great, but that would not be boxable...

I'm not sure the right place where to put this actually... 


What do you think?

Reply:
By adding some indirection it would be possible to have it boxable, but it does not really look to good either https://gist.github.com/fdb-hiroshima/73eca3fd7e4ff47b5c92fad05cbb3a83